types/wlr_wl_shell.c

```diff
@@ -1,61 +1,264 @@
+	}
+
+	event->client = client;
+	event->surface = surface;
+	event->seat_handle = seat_handle;
+	event->serial = serial;
+	event->edges = edges;
+
+	wl_signal_emit(&surface->events.request_resize, event);
+
+	free(event);
 }
 
 static void shell_surface_set_toplevel(struct wl_client *client,
 		struct wl_resource *resource) {
-	wlr_log(L_DEBUG, "TODO: implement shell surface set_toplevel");
+	wlr_log(L_DEBUG, "got shell surface toplevel");
+	struct wlr_wl_shell_surface *surface = wl_resource_get_user_data(resource);
+
+	if (surface->role != WLR_WL_SHELL_SURFACE_ROLE_NONE) {
+		return;
+	}
+
+	surface->role = WLR_WL_SHELL_SURFACE_ROLE_TOPLEVEL;
+
+	wl_signal_emit(&surface->events.set_role, surface);
 }
 
 static void shell_surface_set_transient(struct wl_client *client,
-		struct wl_resource *resource, struct wl_resource *parent, int32_t x,
-		int32_t y, uint32_t flags) {
-	wlr_log(L_DEBUG, "TODO: implement shell surface set_transient");
+		struct wl_resource *resource, struct wl_resource *parent_resource,
+		int32_t x, int32_t y, uint32_t flags) {
+	wlr_log(L_DEBUG, "got shell surface transient");
+	struct wlr_wl_shell_surface *surface = wl_resource_get_user_data(resource);
+	struct wlr_wl_shell_surface *parent =
+		wl_resource_get_user_data(parent_resource);
+	// TODO: check if parent_resource == NULL?
+
+	if (surface->role != WLR_WL_SHELL_SURFACE_ROLE_NONE) {
+		return;
+	}
+
+	struct wlr_wl_shell_surface_transient_state *state =
+		calloc(1, sizeof(struct wlr_wl_shell_surface_transient_state));
+	if (state == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	state->parent = parent;
+	state->x = x;
+	state->y = y;
+	state->flags = flags;
+
+	free(surface->transient_state);
+	surface->transient_state = state;
+
+	surface->role = WLR_WL_SHELL_SURFACE_ROLE_TRANSCIENT;
+
+	wl_signal_emit(&surface->events.set_role, surface);
 }
 
 static void shell_surface_set_fullscreen(struct wl_client *client,
 		struct wl_resource *resource, uint32_t method, uint32_t framerate,
-		struct wl_resource *output) {
-	wlr_log(L_DEBUG, "TODO: implement shell surface set_fullscreen");
+		struct wl_resource *output_resource) {
+	wlr_log(L_DEBUG, "got shell surface fullscreen");
+	struct wlr_wl_shell_surface *surface = wl_resource_get_user_data(resource);
+	struct wlr_output *output = NULL;
+	if (output_resource != NULL) {
+		output = wl_resource_get_user_data(output_resource);
+	}
+
+	if (surface->role == WLR_WL_SHELL_SURFACE_ROLE_TOPLEVEL) {
+		return;
+	}
+
+	struct wlr_wl_shell_surface_set_fullscreen_event *event =
+		calloc(1, sizeof(struct wlr_wl_shell_surface_set_fullscreen_event));
+	if (event == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	event->client = client;
+	event->surface = surface;
+	event->method = method;
+	event->framerate = framerate;
+	event->output = output;
+
+	wl_signal_emit(&surface->events.request_set_fullscreen, event);
+
+	free(event);
 }
 
 static void shell_surface_set_popup(struct wl_client *client,
```

Comment:
We need to do a pointer grab here, but that's not implemented in the seat yet so I don't want to worry about it. Maybe just add a TODO for it.